### PR TITLE
ci: Update artifact actions

### LINKED ISSUE #27189: upgrade artifact upload and dowload actions to v4
### Describe the proposed change(s).

They need manual upgrade because of some breaking changes

### PR DESCRIPTION
**Pull request by zharinov**
## Context

- closes #27189

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by zharinov** [comment]
https://github.com/zharinov/renovate/actions/runs/7889557679

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -366,7 +366,7 @@ jobs:
 
       - name: Save coverage artifacts
         if: (success() || failure()) && github.event.pull_request.draft != true && matrix.coverage
-        uses: actions/upload-artifact@a8a3f3ad30e3422c9c7b888a15615d19a852ae32 # v3.1.3
+        uses: actions/upload-artifact@5d5d22a31266ced268874388b861e4b58bb5c2f3 # v4.3.1
```
Comment (viceice) [inline .github/workflows/build.yml:369]: you need to change the parameters here, checkout the breaking changes

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -385,7 +385,7 @@ jobs:
           sparse-checkout: ${{ env.SPARSE_CHECKOUT }}
 
       - name: Download coverage reports
-        uses: actions/download-artifact@9bc31d5ccc31df68ecc42ccf4149144866c47d8a # v3.0.2
+        uses: actions/download-artifact@eaceaf801fd36c7dee90939fad912460b18a1ffe # v4.1.2
```
Comment (viceice) [inline .github/workflows/build.yml:388]: here too

**Review by viceice — CHANGES_REQUESTED** [review]

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -366,7 +366,7 @@ jobs:
 
       - name: Save coverage artifacts
         if: (success() || failure()) && github.event.pull_request.draft != true && matrix.coverage
-        uses: actions/upload-artifact@a8a3f3ad30e3422c9c7b888a15615d19a852ae32 # v3.1.3
+        uses: actions/upload-artifact@5d5d22a31266ced268874388b861e4b58bb5c2f3 # v4.3.1
         with:
           name: coverage
```
Comment (viceice) [inline .github/workflows/build.yml:371]: you can no longer have artifacts with same name, but download can use wildcards and merge

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -366,9 +366,10 @@ jobs:
 
       - name: Save coverage artifacts
         if: (success() || failure()) && github.event.pull_request.draft != true && matrix.coverage
-        uses: actions/upload-artifact@a8a3f3ad30e3422c9c7b888a15615d19a852ae32 # v3.1.3
+        uses: actions/upload-artifact@5d5d22a31266ced268874388b861e4b58bb5c2f3 # v4.3.1
         with:
-          name: coverage
+          overwrite: true
```
Comment (viceice) [inline .github/workflows/build.yml:371]: this shouldn't needed

**Review by viceice — DISMISSED** [review]

**Comment by viceice** [comment]
🤔 codecov isn't reporting, there's an error. see action

**Comment by zharinov** [comment]
> 🤔 codecov isn't reporting, there's an error. see action

- Introduced in #27236

**Comment by zharinov** [comment]
Okay, let's go with artifact actions only here. Codecov is the separate problem.

**Review by viceice — APPROVED** [review]

**Pull request merged**